qiskit_ibm/runtime/runtime_job.py

```diff
@@ -203,15 +199,18 @@ def wait_for_final_state(
             JobTimeoutError: If the job does not reach a final state before the
                 specified timeout.
         """
+        if not self._is_streaming():
+            self._ws_client_future = self._executor.submit(self._start_websocket_client)
```

Comment:
So if the job is already done (i.e. `self._ws_client_future.done()` is true), we'll still open a connection? Presumably the server will just close it immediate?

Also if I call this method, which times out, then I try to do `stream_results()` (maybe trying to figure out why the job is taking so long), it's going to tell me a callback function is already streaming results, which is not true.

Reply:
I suppose we could do if `self._ws_client_future is None` we can open a connection and avoid the other case 

As for if `wait_for_final_state()` times out and then `stream_results()` is called, is there a better way to handle this other than just updating the error message?

Reply:
> I suppose we could do if `self._ws_client_future is None` we can open a connection and avoid the other case

Except we also have the `cancel_result_streaming()` method which would cancel the future without setting it to `None`. Instead of relying on the status of the future, we can just check if status is in `JOB_FINAL_STATES` and only open a connection if it isn't.

> As for if wait_for_final_state() times out and then stream_results() is called, is there a better way to handle this other than just updating the error message?

Yeah you can update `stream_results()` to not raise immediately if there is a stream already open. You can instead start calling the user's callback function with message from the stream.